File: packages/cli/src/reporting/summary-printer.ts
import chalk from 'chalk';
import { ScanResult } from './json-reporter';

export interface SummaryOptions {
  fixPreview?: boolean;
  confidence?: number;
}

export function printSummary(results: ScanResult, options: SummaryOptions = {}): void {
  console.log('\n' + chalk.bold.blue('═══════════════════════════════════════════════════════════'));
  console.log(chalk.bold.blue('                    GasGuard Scan Report'));
  console.log(chalk.bold.blue('═══════════════════════════════════════════════════════════\n'));

  // Scan metadata
  console.log(chalk.gray(`Scan Path: ${results.scanPath}`));
  console.log(chalk.gray(`Timestamp: ${results.timestamp}`));
  console.log(chalk.gray(`Files Scanned: ${results.scannedFiles}/${results.totalFiles}\n`));

  // Summary statistics
  console.log(chalk.bold('Summary Statistics:'));
  console.log(`  Total Violations: ${chalk.yellow(results.summary.totalViolations.toString())}`);
  console.log(`  Total Gas Savings: ${chalk.green(formatGasSavings(results.summary.totalGasSavings))}\n`);

  // By severity
  console.log(chalk.bold('Violations by Severity:'));
  printSeverity('Critical', results.summary.bySeverity.critical, 'red');
  printSeverity('High', results.summary.bySeverity.high, 'yellow');
  printSeverity('Medium', results.summary.bySeverity.medium, 'yellow');
  printSeverity('Low', results.summary.bySeverity.low, 'blue');
  printSeverity('Info', results.summary.bySeverity.info, 'gray');
  console.log();

  // By rule
  if (Object.keys(results.summary.byRule).length > 0) {
    console.log(chalk.bold('Violations by Rule:'));
    const sortedRules = Object.entries(results.summary.byRule)
      .sort(([, a], [, b]) => b - a)
      .slice(0, 10); // Show top 10
    
    for (const [rule, count] of sortedRules) {
      console.log(`  ${chalk.cyan(rule)}: ${count}`);
    }
    console.log();
  }

  // Findings details
  if (results.findings.length > 0) {
    console.log(chalk.bold('Findings Details:'));
    console.log(chalk.gray('─'.repeat(60)));
    
    for (const finding of results.findings.slice(0, 20)) { // Show first 20
      printFinding(finding, options);
    }
    
    if (results.findings.length > 20) {
      console.log(chalk.gray(`\n... and ${results.findings.length - 20} more findings`));
    }
    console.log();
  }

  console.log(chalk.bold.blue('═══════════════════════════════════════════════════════════\n'));
}

function printSeverity(label: string, count: number, color: string): void {
  const coloredCount = count > 0 ? chalk[color](count.toString()) : chalk.gray('0');
  console.log(`  ${label.padEnd(10)}: ${coloredCount}`);
}

function printFinding(finding: any, options: SummaryOptions): void {
  const severityColor = getSeverityColor(finding.severity);
  const confidence = finding.confidence ? ` (confidence: ${finding.confidence.toFixed(2)})` : '';
  
  console.log(`\n${chalk.bold(severityColor(`[${finding.severity.toUpperCase()}]`))} ${finding.ruleId} - ${finding.ruleName}`);
  console.log(chalk.gray(`  File: ${finding.file}:${finding.line}`));
  console.log(chalk.gray(`  Message: ${finding.message}`));
  
  if (finding.gasSavings) {
    console.log(chalk.green(`  Gas Savings: ${formatGasSavings(finding.gasSavings)}`));
  }
  
  if (confidence) {
    console.log(chalk.gray(confidence));
  }
  
  if (options.fixPreview && finding.suggestion) {
    console.log(chalk.cyan(`  Suggestion: ${finding.suggestion}`));
  }
}

function getSeverityColor(severity: string): (text: string) => string {
  switch (severity.toLowerCase()) {
    case 'critical':
    case 'error':
      return chalk.red;
    case 'high':
      return chalk.yellow;
    case 'medium':
      return chalk.yellow;
    case 'low':
      return chalk.blue;
    case 'info':
    default:
      return chalk.gray;
  }
}

function formatGasSavings(gas: number): string {
  if (gas >= 1000000) {
    return `${(gas / 1000000).toFixed(2)}M gas`;
  } else if (gas >= 1000) {
    return `${(gas / 1000).toFixed(2)}K gas`;
  }
  return `${gas} gas`;
}

export function printCompactSummary(results: ScanResult): void {
  const critical = results.summary.bySeverity.critical;
  const high = results.summary.bySeverity.high;
  const medium = results.summary.bySeverity.medium;
  const low = results.summary.bySeverity.low;
  
  let status = chalk.green('✓ PASS');
  if (critical > 0) {
    status = chalk.red('✗ FAIL');
  } else if (high > 0) {
    status = chalk.yellow('⚠ WARN');
  }
  
  console.log(`${status} ${results.summary.totalViolations} issues found (${critical} critical, ${high} high, ${medium} medium, ${low} low) - ${formatGasSavings(results.summary.totalGasSavings)} potential savings`);
}

File: src/reporting/summary_printer.rs
use analysis_core::plugin::Finding;
use std::io::{self, Write};

pub struct SummaryPrinter {
    use_colors: bool,
}

impl SummaryPrinter {
    pub fn new() -> Self {
        Self { use_colors: true }
    }

    pub fn with_colors(use_colors: bool) -> Self {
        Self { use_colors }
    }

    pub fn print_summary(&self, findings: &[Finding], scan_path: &str) {
        self.print_header();
        self.print_metadata(scan_path, findings.len());
        self.print_statistics(findings);
        self.print_findings(findings);
        self.print_footer();
    }

    pub fn print_compact(&self, findings: &[Finding]) {
        let critical = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Critical))
            .count();
        let high = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Error))
            .count();
        let medium = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Warning))
            .count();
        let low = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Info))
            .count();

        let status = if critical > 0 {
            self.color("✗ FAIL", "red")
        } else if high > 0 {
            self.color("⚠ WARN", "yellow")
        } else {
            self.color("✓ PASS", "green")
        };

        println!(
            "{} {} issues found ({} critical, {} high, {} medium, {} low)",
            status,
            findings.len(),
            critical,
            high,
            medium,
            low
        );
    }

    fn print_header(&self) {
        println!("{}", self.color("═══════════════════════════════════════════════════════════", "blue"));
        println!("{}", self.color("                    GasGuard Scan Report", "blue"));
        println!("{}\n", self.color("═══════════════════════════════════════════════════════════", "blue"));
    }

    fn print_metadata(&self, scan_path: &str, count: usize) {
        println!("Scan Path: {}", self.color(scan_path, "gray"));
        println!("Files Scanned: {}\n", self.color(&count.to_string(), "gray"));
    }

    fn print_statistics(&self, findings: &[Finding]) {
        println!("{}", self.color("Summary Statistics:", "bold"));
        println!("  Total Violations: {}", self.color(&findings.len().to_string(), "yellow"));
        
        let critical = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Critical))
            .count();
        let high = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Error))
            .count();
        let medium = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Warning))
            .count();
        let low = findings.iter()
            .filter(|f| matches!(f.severity, analysis_core::plugin::Severity::Info))
            .count();

        println!("\n{}", self.color("Violations by Severity:", "bold"));
        self.print_severity_count("Critical", critical, "red");
        self.print_severity_count("High", high, "yellow");
        self.print_severity_count("Medium", medium, "yellow");
        self.print_severity_count("Low", low, "blue");
        println!();
    }

    fn print_severity_count(&self, label: &str, count: usize, color: &str) {
        let colored_count = if count > 0 {
            self.color(&count.to_string(), color)
        } else {
            self.color("0", "gray")
        };
        println!("  {:10}: {}", label, colored_count);
    }

    fn print_findings(&self, findings: &[Finding]) {
        if findings.is_empty() {
            return;
        }

        println!("{}", self.color("Findings Details:", "bold"));
        println!("{}", self.color("─".repeat(60), "gray"));

        for finding in findings.iter().take(20) {
            self.print_finding(finding);
        }

        if findings.len() > 20 {
            println!("\n... and {} more findings", findings.len() - 20);
        }
        println!();
    }

    fn print_finding(&self, finding: &Finding) {
        let severity_str = format!("{:?}", finding.severity);
        let severity_color = match finding.severity {
            analysis_core::plugin::Severity::Critical => "red",
            analysis_core::plugin::Severity::Error => "yellow",
            analysis_core::plugin::Severity::Warning => "yellow",
            analysis_core::plugin::Severity::Info => "blue",
        };

        println!("\n[{}] {} - {}",
            self.color(&severity_str, severity_color),
            finding.rule_id,
            finding.rule_id
        );
        println!("  File: {}:{}", finding.file, finding.line);
        println!("  Message: {}", self.color(&finding.message, "gray"));
        
        if let Some(ref suggestion) = finding.suggestion {
            println!("  Suggestion: {}", self.color(suggestion, "cyan"));
        }
    }

    fn print_footer(&self) {
        println!("{}\n", self.color("═══════════════════════════════════════════════════════════", "blue"));
    }

    fn color(&self, text: &str, color: &str) -> String {
        if !self.use_colors {
            return text.to_string();
        }

        match color {
            "red" => format!("\x1b[31m{}\x1b[0m", text),
            "green" => format!("\x1b[32m{}\x1b[0m", text),
            "yellow" => format!("\x1b[33m{}\x1b[0m", text),
            "blue" => format!("\x1b[34m{}\x1b[0m", text),
            "cyan" => format!("\x1b[36m{}\x1b[0m", text),
            "gray" => format!("\x1b[90m{}\x1b[0m", text),
            "bold" => format!("\x1b[1m{}\x1b[0m", text),
            _ => text.to_string(),
        }
    }
}

impl Default for SummaryPrinter {
    fn default() -> Self {
        Self::new()
    }
}
